### ISSUE TITLE
Does CLI support initialisation?

### ISSUE THREAD

**Issue by leinelissen** [bailey27/cppcryptfs#108]
Hi there!

First of all, thanks for your outstanding work in the open source community. I would love to incorporate your library into software I am building for personal data management. I have full compatibility with GocryptFS at this point, but I am running into a compatibility issue. During application startup, I check if the filesystem is initialised, and if not, I create it. This can be done easily on *nix by doing `gocryptfs -init path`, yet I cannot find an equivalent CLI option to initialise a file system with cppcryptfs.

Am I missing something, or is it perhaps trivial to add?

Cheers!
Lei

**Comment by bailey27** [comment]
Currently, cppcryptfs can create a new filesystem only by using the GUI.

I could add the capability for creating one from the command line.

How do you prompt for the password when you run "gocryptfs -init" from your program?

Would throwing up a console window for this be OK for you?

The way I think it would work would be for cppcryptfsctl (the companion console-mode program) to be run as 

cppcryptfsctl --init some_path [args]

I assume your app is a gui app, so if it ran cppcryptfsctl, that would cause Windows to create and display a console window to run cppcryptfsctl in, and it would prompt for the password in the console window.

cppcryptfs is a GUI app, and interacting with a console is difficult from a Windows GUI app and seems to have limitations (e.g. output redirection doesn't work).  So that's why I think it would be better to use cppcryptfsctl for this.

**Comment by leinelissen** [comment]
Currently, gocryptfs reads passwords from stdin. I wait for it to prompt for a password, and then I pass it directly to stdin, as described [here](https://github.com/leinelissen/aeon/blob/918a8018effdee92cc74471bea8c9c2c7fe12ed7/src/main/lib/crypto-fs/index.ts#L119). 

Your assumption that I am building a GUI app is correct. I am building an Electron-based app, in which gocryptfs is called from Node by means of a child process. As this is managed via a child process, I currently have no need for spawning console windows, and I would assume this would not be necessary for cppcryptsfs either (please let me know if this is wrong though). 

It does indeed seem like cppcryptfsctl would be more suited in this case. I'll take your guidance on for which part of cppcryptfs it is more logical to implement. It's suffices to say that if it's not too much work for you to implement, it'd be awesome if you helped out!

**Comment by bailey27** [comment]
I got it working pretty much.  cppcryptfsctl can make filesystems now.

It does the Windows equivalent of isatty() on stdin, and if stdin is a tty, it prompts twice (password and repeat password) without echo.

If stdin isn't a tty, then it reads the password from stdin.

I hope to release it this week or maybe this coming weekend.

**Comment by leinelissen** [comment]
That's amazing! Thanks so much for helping out, I am looking forward to testing it.

**Comment by bailey27** [comment]
I'm attaching a version in which cppcryptfsctl supports initialization.  It has "INIT-TEST" in the version string.

[cppcryptfs_cppcryptfsctl_cmd_init_test.zip](https://github.com/bailey27/cppcryptfs/files/5115009/cppcryptfs_cppcryptfsctl_cmd_init_test.zip)

I had to get a new code singing certificate because my old one was about to expire. 

You'll probably have to click on "more info" and "run anyway" because my new cert hasn't had time to earn a reputation with Microsoft.

I want to point out that only cppcryptfsclt supports initialization from the command line (not cppcryptfs).

Normally, cppcryptfsctrl tries to connect to a running instance if cppcryptfs, and if it can, then it sends the command line to it and cppcryptfs actually executes the command line.

Now, when cppcryptfsctl sees the --init option in the command line, it handles the command line itself.  

It should be pretty transparent to how everything was operating before, but I just wanted to point that out.

cppcryptfsctl will read the password from stdin if stdin is not a tty.  If stdin is a tty, then it will prompt for the password and repeat password.

Here is what the command line help looks like now.

```
Usage: cppcryptfs/cppcryptfsctl [OPTIONS]

Mounting:
  -m, --mount=PATH         mount filesystem located at PATH
  -d, --drive=D            mount to drive letter D or empty dir DIR
  -p, --password=PASS      use password PASS
  -P, --saved-password     use saved password
  -r, --readonly           mount read-only
  -c, --config=PATH        path to config file for init/mount
  -s, --reverse            init/mount reverse filesystem (implies siv for init)

Unmounting:
  -u, --unmount=D          unmount drive letter D or dir DIR
  -u, --unmount=all        unmount all drives

Misc:
  -t, --tray               hide in system tray
  -x, --exit               exit if no drives mounted
  -l, --list               list available and mounted drive letters (with paths)
  -ld:\p, --list=d:\p      list plaintext and encrypted filenames
  -C, --csv                file list is comma-delimited
  -D, --dir                file list dirs first and w/ trailing \"\\\"
  -i, --info=D             show information about mounted filesystem
  -v, --version            print version (use --init -v for cppcryptfsctl ver)
  -h, --help               display this help message

Initializing (cppcryptfsctl only):
  -I, --init=PATH          Initialize encrypted filesystem located at PATH
  -V, --volumename=NAME    specify volume name for filesystem
  -T, --plaintext          use plaintext filenames (default is AES256-EME)
  -S, --siv                use AES256-SIV for data encryption (default is GCM)
  -L, --longnames [0|1]    enable/disable LFNs. defaults to enabled (1)
  -b, --streams   [0|1]    enable/disable streams. defaults to enabled (1)
```

The arguments that are common to mounting and initializing also work (e.g. --config and --reverse).

Please let me know if you need me to change anything or if you need 32-bit binaries.

If you have the annoying problem with Windows opening a transient console window, I can see if I can do anything about that from my end.   Electron might already prevent that. Let's hope.

**Comment by leinelissen** [comment]
That does the trick! Fortunately, NodeJS ends up surpressing terminal Windows, so the process is wonderfully smooth for the end-user. Thanks again for your time on this, I truly appreciate it.

**Comment by bailey27** [comment]
That's great news!  

I plan to release a new version with this code on the releases page in the next few days.

**Comment by bailey27** [comment]
I've released version 1.4.3.1 with this code.

**Comment by leinelissen** [comment]
Again, that's amazing. Thanks so much for your help!
